### PR TITLE
[CORE-346] Add getGoogleProjects and getGoogleProjectById APIs

### PR DESCRIPTION
**Pull request by kevinmarete**
Ticket: https://broadworkbench.atlassian.net/browse/CORE-346

This PR adds two endpoints to manage Google projects:

- `GET /googleProjects` Lists Google projects the user can access. It also supports optional filtering by billing project.
- `GET /googleProjects/{googleProjectId}` Retrieves a Google project if accessible.

#### Notes:
---
- Includes error handling for missing or unauthorized projects.
- Tests cover both endpoints and edge cases.
- [Added](https://github.com/broadinstitute/sam/pull/1656) `read` action to sam to support permissions check

---

**PR checklist**

- [x] Include the JIRA issue number in the PR description and title
- [x] Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] If you changed anything in `model/`, then you should publish a new official `rawls-model` and perform the corresponding dependency updates as specified in the [README](https://github.com/broadinstitute/rawls/blob/develop/README.md#publish-rawls-model):
  - [ ] in the automation subdirectory
  - [ ] in workbench-libs
  - [ ] in firecloud-orchestration
- [x] Get two thumbsworth of PR review
- [x] Verify all tests go green, including CI tests
- [x] **Squash commits and merge** to develop (branches are automatically deleted after merging)
- [ ] Inform other teams of any substantial changes via Slack and/or email

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala [hunk 1]
```diff
@@ -109,4 +112,44 @@ class GoogleProjectRegistrationService(protected val ctx: RawlsRequestContext,
         case None => Future.successful(None)
       }
     } yield finalResult
+
+  def getGoogleProjects(billingProjectName: Option[RawlsBillingProjectName]): Future[Seq[GoogleProjectRegistration]] = {
+    val accessibleGoogleProjectsFuture = samDAO
+      .listResourcesWithActions(SamResourceTypeNames.googleProject, SamGoogleProjectActions.readPolicies, ctx)
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala]: `listResourcesWithActions` will list only google projects that the user has `readPolicies` on which is currently only owner. The JIRA says "List Google projects that the user has access to." What the code does and what the JIRA says are different but I am not sure which we actually want. Who is using this api and for what?

If we want to limit responses to just projects the user owns, `readPolicies` is probably the wrong action to check anyway. My suspicion is that we want to use the `link` action.

[Use this api to do what the JIRA says](https://github.com/broadinstitute/rawls/blob/463209c785e252a3cc8b197853459318104ef9c3/core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpSamDAO.scala#L241).

File: core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala [hunk 2]
```diff
@@ -109,4 +112,44 @@ class GoogleProjectRegistrationService(protected val ctx: RawlsRequestContext,
...
+      .map(_.map(resource => GoogleProjectId(resource.getResourceId)).toSet)
+
+    accessibleGoogleProjectsFuture.flatMap { accessibleGoogleProjects =>
+      googleProjectRegRepo.getGoogleProjectRegistrations(accessibleGoogleProjects).map { registrations =>
+        val filteredRegistrations = billingProjectName
+          .map(name => registrations.filter(_.billingProjectId == name))
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala]: Do this filtering in the db query

File: core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala [hunk 3]
```diff
@@ -109,4 +112,44 @@ class GoogleProjectRegistrationService(protected val ctx: RawlsRequestContext,
...
+        }
+      }
+    }
+  }
+
+  def getGoogleProjectById(googleProjectId: GoogleProjectId): Future[Option[GoogleProjectRegistration]] =
+    samDAO
+      .userHasAction(SamResourceTypeNames.googleProject,
+                     googleProjectId.value,
+                     SamGoogleProjectActions.readPolicies,
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala]: readPolicies does not seem like the right action to check

File: core/src/main/resources/swagger/api-docs.yaml [hunk 4]
```diff
@@ -4658,6 +4658,37 @@ paths:
...
+      description: List the Google projects accessible to the user
+      operationId: getGoogleProjects
+      parameters:
+        - name: billingProjectId
+          in: query
+          description: |
+            Optional. If provided, only return google projects that are linked to the given billing project.
+          required: false
+          schema:
+            type: string
```
Comment (davidangb) [inline core/src/main/resources/swagger/api-docs.yaml:4674]: we talked about this in refinement, but I don't see anything in the ticket requirements: this API should probably support pagination. I can imagine data repo having enough projects that this API becomes problematic for some users if not paginated.

File: core/src/main/resources/swagger/api-docs.yaml [hunk 5]
```diff
@@ -4690,6 +4721,30 @@ paths:
...
+      operationId: getGoogleProjectById
+      parameters:
+        - $ref: '#/components/parameters/googleProjectIdPathParam'
+      responses:
+        200:
+          description: Successful Request
+          content:
+            'application/json':
+              schema:
+                $ref: '#/components/schemas/RegisterGoogleProjectRequest'
```
Comment (davidangb) [inline core/src/main/resources/swagger/api-docs.yaml]: for both of these APIs, I think we should also return at least the `message` on the project (if it exists). We may also want to return the billing account - not sure about that one; returning the billing project id might be enough.

File: core/src/main/resources/swagger/api-docs.yaml [hunk 6]
```diff
@@ -7463,6 +7518,13 @@ components:
       required: true
       schema:
         type: string
+    googleProjectIdPathParam:
+      name: googleProjectId
+      in: path
+      description: Google Project ID
+      required: true
+      schema:
+        type: string
```
Comment (davidangb) [inline core/src/main/resources/swagger/api-docs.yaml]: will need to coordinate with #3225 on this; see also my comment on that other PR:

https://github.com/broadinstitute/rawls/pull/3225/files#r2006421767

File: core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala (hunk 1 above)
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala]: why check the `read_policies` action? I think this should just check a simple `read` action. Unfortunately I don't see that already in Sam; you would have to add it.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala [hunk 7]
```diff
@@ -109,4 +112,44 @@ class GoogleProjectRegistrationService(protected val ctx: RawlsRequestContext,
...
+  def getGoogleProjects(billingProjectName: Option[RawlsBillingProjectName]): Future[Seq[GoogleProjectRegistration]] = {
+    val accessibleGoogleProjectsFuture = samDAO
+      .listResourcesWithActions(SamResourceTypeNames.googleProject, SamGoogleProjectActions.readPolicies, ctx)
+      .map(_.map(resource => GoogleProjectId(resource.getResourceId)).toSet)
+
+    accessibleGoogleProjectsFuture.flatMap { accessibleGoogleProjects =>
+      googleProjectRegRepo.getGoogleProjectRegistrations(accessibleGoogleProjects).map { registrations =>
+        val filteredRegistrations = billingProjectName
+          .map(name => registrations.filter(_.billingProjectId == name))
+          .getOrElse(registrations)
```
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala]: this queries the db to return all `google-project` rows the user can see, then immediately discards all google-projects which don't match the requested billing project. Instead, could you include the billing project in the db query where clause, so the db only returns the relevant rows?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala [hunk 8]
```diff
@@ -109,4 +112,44 @@ class GoogleProjectRegistrationService(protected val ctx: RawlsRequestContext,
...
+
+        if (billingProjectName.isDefined && filteredRegistrations.isEmpty) {
+          Seq.empty
+        } else {
+          val registeredProjectIds = filteredRegistrations.map(_.googleProjectId).toSet
+          val unregisteredProjects =
+            accessibleGoogleProjects.diff(registeredProjectIds).map(UnRegisteredGoogleProjectRegistration.apply)
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala]: Returning a magic string of "UNREGISTERED" is not a great practice for an api. It might be good for a human to read, but how can a general api consumer know that "UNREGISTERED" isn't the actual name of a google project?

I suggest making `billingProjectId` an Option and returning `None` for the unregistered case. If we want to be more explicit, we can also have a boolean field:
```
val registered: Boolean = billingProjectId.isEmpty
```

This change along with doing more filtering in the database should simplify this function to 2 steps
1. list the accessible projects from sam
2. look them up in the database
These conditionals, etc. will go away

File: core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/GoogleProjectRegistrationApiService.scala [hunk 9]
```diff
@@ -20,22 +25,45 @@ trait GoogleProjectRegistrationApiService extends UserInfoDirectives {
...
+            }
+        } ~
+          path(Segment) { googleProjectId =>
+            get {
+              onSuccess(
+                googleProjectRegServiceConstructor(ctx)
+                  .getGoogleProjectById(GoogleProjectId(googleProjectId))
+              ) {
+                case Some(projectId) => complete(StatusCodes.OK -> projectId)
+                case None            => complete(StatusCodes.NotFound)
```
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/GoogleProjectRegistrationApiService.scala]: Both these APIs should use the Rawls-standard "does not exist or you do not have permission to use it" message. To make this happen you probably will need to throw an error from `GoogleProjectRegistrationService` instead of handling the Some/None cases here.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala [hunk 10]
```diff
@@ -109,4 +112,44 @@ class GoogleProjectRegistrationService(protected val ctx: RawlsRequestContext,
...
+                     SamGoogleProjectActions.readPolicies,
+                     ctx
+      )
+      .flatMap {
+        case true =>
+          googleProjectRegRepo.getGoogleProjectRegistration(googleProjectId).map {
+            case Some(project) => Some(project)
+            case None          => Some(UnRegisteredGoogleProjectRegistration(googleProjectId))
+          }
+        case false => Future.successful(None)
```
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala]: if the user does not have permission to read the requested project, or if there is no row in the db for the requested project, throw a `RawlsExceptionWithErrorReport` with an `ErrorReport` specifying status code 404 and the Rawls-standard "does not exist or you do not have permission to use it" message. This will also simplify handling in your ApiService layer; the api layer then doesn't need to handle Some vs. None

File: core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/GoogleProjectRegistrationApiService.scala [hunk 11]
```diff
@@ -20,22 +25,45 @@ trait GoogleProjectRegistrationApiService extends UserInfoDirectives {
...
+              }
+            }
+          } ~
+            get {
+              parameters("billingProjectId".optional) { billingProjectId =>
+                onSuccess(
+                  googleProjectRegServiceConstructor(ctx)
+                    .getGoogleProjects(billingProjectId.map(RawlsBillingProjectName))
+                ) { projects =>
+                  if (projects.isEmpty) complete(StatusCodes.NotFound -> None)
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/GoogleProjectRegistrationApiService.scala]: An empty response should not return a 404. It is fine to return an empty list.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala (hunk 8 above)
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala]: is it even possible to reach this case? If so, there may be a logic error somewhere. The billing-project column in the db is not nullable, so I wouldn't expect to find this at all.

I'd love to get rid of the magic "UNREGISTERED" string and the `UnRegisteredGoogleProjectRegistration` helper entirely; I think we can find more standard solutions for that

**Review by davidangb — CHANGES_REQUESTED** [review]
Initial comments inline. I haven't reviewed tests yet; will do so after the PR is revised. Major review points:
* consider pagination for the list API
* use `read` action instead of `read_policies`
* perform filtering in the db, not in Scala
* don't use a magic string "UNREGISTERED"
* respond with the "does not exist or you do not have permission to use it" message when appropriate

**Review by dvoet — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala (hunk 1 above)
Comment (davidangb) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala]: +1 to this comment. I know I suggested using a `read` action … is `link` the one we really want? In terms of requirements: should this list all google-projects the user can see, even if the user can't assign billing to that google-project? Or should this only list those google-projects for which the user can control billing?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/GoogleProjectRegistrationApiService.scala [hunk 12]
```diff
@@ -35,7 +40,19 @@ trait GoogleProjectRegistrationApiService extends UserInfoDirectives {
                   }
               }
             }
-          }
+          } ~
+            get {
+              parameters(
+                "billingProjectId".optional,
+                "pageSize".as[Int],
+                "offset".as[Int]
```
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/GoogleProjectRegistrationApiService.scala]: `pageSize` and `offset` are defined as optional in the OpenAPI spec, and therefore callers may not specify them. But when omitted, this API throws an error. Give these default values here by using `.withDefault(n)` so they are actually optional.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala [hunk 13]
```diff
@@ -127,4 +128,30 @@ class GoogleProjectRegistrationService(protected val ctx: RawlsRequestContext,
...
+  }
+
+  def getGoogleProjectById(googleProjectId: GoogleProjectId): Future[Option[GoogleProjectRegistration]] =
+    for {
+      _ <- samDAO
+        .userHasAction(SamResourceTypeNames.googleProject, googleProjectId.value, SamGoogleProjectActions.read, ctx)
+        .map(canRead =>
+          if (!canRead)
+            throw new RawlsExceptionWithErrorReport(errorReport =
+              ErrorReport(StatusCodes.Forbidden, s"Google project not found or you do not have permission to read.")
```
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala]: nit: don't specify the exact action here (because we might change it in the future). Generally in Rawls we use "… you do not have permission to use it" or "… you do not have access".

File: core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala [hunk 14]
```diff
@@ -127,4 +128,30 @@ class GoogleProjectRegistrationService(protected val ctx: RawlsRequestContext,
       // 3. Delete the record in the GOOGLE_PROJECT table.
       _ <- googleProjectRegRepo.deleteGoogleProjectRegistration(googleProjectId)
     } yield ()
+
+  def getGoogleProjects(billingProjectName: Option[RawlsBillingProjectName],
+                        pageSize: Int,
+                        offset: Int
+  ): Future[Seq[GoogleProjectRegistration]] = {
```
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala:135]: consider adding some validation here: `offset` is >=0; `pageSize` is a sane number, say `0 < pageSize < 5000` or wherever we think we should put a hard limit on the response size

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/GoogleProjectRegistrationComponentSpec.scala [hunk 15]
```diff
@@ -53,4 +53,41 @@ class GoogleProjectRegistrationComponentSpec
...
+                                                                ),
+                                                                None,
+                                                                10,
+                                                                0
+      )
+    )
+
+    result should have size 2
+    result.map(_.googleProjectId.value) should contain allOf ("project1", "project2")
+  }
```
Comment (davidangb) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/GoogleProjectRegistrationComponentSpec.scala:154]: consider adding a test for pagination, here or in the ServiceSpec (probably don't need both)

File: core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala (hunk 1 above)
Comment (kevinmarete) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala]: As agreed in standup parking lot we will use new `read` action. https://github.com/broadinstitute/sam/pull/1656

File: core/src/test/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationRepositorySpec.scala [hunk 16]
```diff
@@ -86,4 +86,59 @@ class GoogleProjectRegistrationRepositorySpec extends AnyFlatSpec with TestDrive
...
+    val repo = new GoogleProjectRegistrationRepository(slickDataSource)
+    val googleProjectReg = makeGoogleProjectRegistration(testData.testProject1)
+
+    Await.result(repo.registerGoogleProject(googleProjectReg), Duration.Inf)
+    val result = Await.result(repo.getGoogleProjectRegistration(googleProjectReg.googleProjectId), Duration.Inf)
+
+    result match {
+      case Some(project) => assertResult(googleProjectReg)(project)
+      case None          => fail(s"Expected Some(${googleProjectReg}) but got None")
+    }
```
Comment (davidangb) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationRepositorySpec.scala]: totally optional; this is for your learning the codebase and frameworks. If this test class added the `org.scalatest.matchers.should.Matchers` trait (`class GoogleProjectRegistrationRepositorySpec extends AnyFlatSpec with TestDriverComponent with Matchers`), then you could write this using the [matchers DSL](https://www.scalatest.org/user_guide/using_matchers), which can be terser and arguably more readable:
```scala
result should contain(googleProjectReg)
```

I especially note this is optional feedback because you may not want to mix assertion styles within the same class.

Edited to add: I see you using Matchers in other tests! So you probably know all this already.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationRepositorySpec.scala [hunk 17]
```diff
@@ -86,4 +86,59 @@ class GoogleProjectRegistrationRepositorySpec extends AnyFlatSpec with TestDrive
...
+
+  it should "return None if google project registration id does not exist" in withDefaultTestDatabase {
+    val repo = new GoogleProjectRegistrationRepository(slickDataSource)
+    val nonExistentId = GoogleProjectId(UUID.randomUUID().toString)
+
+    val result = Await.result(repo.getGoogleProjectRegistration(nonExistentId), Duration.Inf)
+
+    assertResult(None) {
+      result
+    }
```
Comment (davidangb) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationRepositorySpec.scala]: using Matchers:
```
result shouldBe empty
```

File: core/src/test/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationRepositorySpec.scala [hunk 18]
```diff
@@ -86,4 +86,59 @@ class GoogleProjectRegistrationRepositorySpec extends AnyFlatSpec with TestDrive
...
+      repo.getGoogleProjectRegistrations(Set(googleProjectReg1.googleProjectId, googleProjectReg2.googleProjectId),
+                                         None,
+                                         10,
+                                         0
+      ),
+      Duration.Inf
+    )
+
+    assert(result.contains(googleProjectReg1))
+    assert(result.contains(googleProjectReg2))
```
Comment (davidangb) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationRepositorySpec.scala]: using Matchers:
```
result should contain theSameElementsAs List(googleProjectReg1, googleProjectReg2)
```
in this case, the Matcher DSL is more precise. The existing code would still pass if the result was `List(googleProjectReg1, googleProjectReg2, googleProjectReg3, googleProjectReg4)`. You could also solve this by adding `assertResult(2)(result.size)` and not using Matchers.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationRepositorySpec.scala [hunk 19]
```diff
@@ -86,4 +86,59 @@ class GoogleProjectRegistrationRepositorySpec extends AnyFlatSpec with TestDrive
...
+
+  it should "return an empty sequence if none of the google project registration ids exist" in withDefaultTestDatabase {
+    val repo = new GoogleProjectRegistrationRepository(slickDataSource)
+    val nonExistentId1 = GoogleProjectId(UUID.randomUUID().toString)
+    val nonExistentId2 = GoogleProjectId(UUID.randomUUID().toString)
+
+    val result =
+      Await.result(repo.getGoogleProjectRegistrations(Set(nonExistentId1, nonExistentId2), None, 10, 0), Duration.Inf)
+
+    assert(result.isEmpty)
```
Comment (davidangb) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationRepositorySpec.scala]: using Matchers:
```
result shouldBe empty
```

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/GoogleProjectRegistrationComponent.scala [hunk 20]
```diff
@@ -73,6 +72,22 @@ trait GoogleProjectRegistrationComponent {
           DBIO.successful(Option(GoogleProjectRegistrationRecord.toGoogleProjectRegistration(googleProjectRegRec)))
       }
 
+    def findByIdsAndBillingProject(
+      googleProjectIds: Set[GoogleProjectId],
+      billingProjectName: Option[RawlsBillingProjectName],
+      pageSize: Int,
+      offset: Int
+    ): ReadAction[Seq[GoogleProjectRegistration]] = {
+      val query = googleProjectRegistrationQuery.filter(_.googleProjectId inSet googleProjectIds.map(_.value))
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/GoogleProjectRegistrationComponent.scala]: What happens when `googleProjectIds` is empty?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/GoogleProjectRegistrationComponent.scala [hunk 21]
```diff
@@ -73,6 +72,22 @@ trait GoogleProjectRegistrationComponent {
...
+      val filteredQuery = billingProjectName match {
+        case Some(name) => query.filter(_.billingProject === name.value)
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/GoogleProjectRegistrationComponent.scala]: maybe a little nicer, up to you
```suggestion
        case Some(RawlsBillingProjectName(name)) => query.filter(_.billingProject === name)
```

File: core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala [hunk 22]
```diff
@@ -127,4 +128,30 @@ class GoogleProjectRegistrationService(protected val ctx: RawlsRequestContext,
...
+    val accessibleGoogleProjectsFuture = samDAO
+      .listResourcesWithActions(SamResourceTypeNames.googleProject, SamGoogleProjectActions.read, ctx)
+      .map(_.map(resource => GoogleProjectId(resource.getResourceId)).toSet)
+
+    accessibleGoogleProjectsFuture.flatMap { accessibleGoogleProjects =>
+      googleProjectRegRepo.getGoogleProjectRegistrations(accessibleGoogleProjects, billingProjectName, pageSize, offset)
+    }
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala]: maybe a little nicer as a for comprehension, your choice
```suggestion
    for {
      accessibleGoogleProjectResources <- samDAO
        .listResourcesWithActions(SamResourceTypeNames.googleProject, SamGoogleProjectActions.read, ctx)

      accessibleGoogleProjects = accessibleGoogleProjectResources.map(resource => GoogleProjectId(resource.getResourceId)).toSet
      
      projectRegistrations <- googleProjectRegRepo.getGoogleProjectRegistrations(accessibleGoogleProjects, billingProjectName, pageSize, offset)
    } yield projectRegistrations
```

File: core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala (hunk 13 above)
Comment (dvoet) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala]: This should probably be a 404 error with a message like "Google project does not exist or you don't have access." We don't want the caller to be able to distinguish between the 2 cases because that leaks information.

Better yet just return `None` at this point.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/GoogleProjectRegistrationComponentSpec.scala [hunk 23]
```diff
@@ -53,4 +53,41 @@ class GoogleProjectRegistrationComponentSpec
 
   }
 
+  it should "findByIDs" in withDefaultTestDatabase {
```
Comment (dvoet) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/GoogleProjectRegistrationComponentSpec.scala]: add a test that passes in an empty set to `findByIdsAndBillingProject`

File: core/src/test/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationRepositorySpec.scala [hunk 24]
```diff
@@ -86,4 +86,59 @@ class GoogleProjectRegistrationRepositorySpec extends AnyFlatSpec with TestDrive
 
   }
 
+  it should "retrieve a google project registration by id" in withDefaultTestDatabase {
```
Comment (dvoet) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationRepositorySpec.scala:83]: The functions added to `GoogleProjectRegistrationRepository` are only pass throughs. The test cases added here should be covered by `GoogleProjectRegistrationComponent` or vice versa. I don't think you need the same coverage in both places. Make sure you have the right coverage in one place or the other but I don't think you need both.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationServiceSpec.scala [hunk 25]
```diff
@@ -692,4 +693,256 @@ class GoogleProjectRegistrationServiceSpec
...
+                                                             0
+      )
+    )
+      .thenReturn(Future.successful(Seq(googleProjectRegistration)))
+
+    val result = Await.result(googleProjectRegService.getGoogleProjects(Some(billingProjectName), 10, 0), Duration.Inf)
+    result shouldEqual Seq(googleProjectRegistration)
+  }
+
+  it should "return NotFound if billing project name is set but no matches are found" in {
```
Comment (dvoet) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationServiceSpec.scala]: test name is incorrect

File: core/src/test/scala/org/broadinstitute/dsde/rawls/model/SamModelSpec.scala [hunk 26]
```diff
@@ -0,0 +1,15 @@
+package org.broadinstitute.dsde.rawls.model
+
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+class SamModelSpec extends AnyFlatSpec with Matchers {
```
Comment (dvoet) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/model/SamModelSpec.scala]: I don't think these tests are worthwhile

File: core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/GoogleProjectRegistrationApiServiceSpec.scala [hunk 27]
```diff
@@ -0,0 +1,119 @@
...
+
+    Get(
+      "/googleProjects?billingProjectId=test-billing-project&pageSize=10&offset=0"
+    ) ~> googleProjectRegistrationRoutes() ~> check {
+      status shouldEqual StatusCodes.OK
+      responseAs[Seq[GoogleProjectRegistration]] shouldEqual Seq(googleProjectRegistration)
+    }
+  }
+
+  it should "return NotFound if billing project name is set but no matches are found" in {
```
Comment (dvoet) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/GoogleProjectRegistrationApiServiceSpec.scala]: this is named incorrectly but I don't think it is actually different than the test above, it just returns what the mock says, so remove it

File: core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/GoogleProjectRegistrationApiServiceSpec.scala [hunk 28]
```diff
@@ -0,0 +1,119 @@
...
+    when(mockGoogleProjectRegService.getGoogleProjects(None, 10, 0))
+      .thenReturn(Future.successful(Seq(googleProjectRegistration1, googleProjectRegistration2)))
+
+    Get("/googleProjects?pageSize=10&offset=0") ~> googleProjectRegistrationRoutes() ~> check {
+      status shouldEqual StatusCodes.OK
+      responseAs[Seq[GoogleProjectRegistration]] shouldEqual Seq(googleProjectRegistration1, googleProjectRegistration2)
+    }
+  }
+
+  it should "retrieve a Google project by ID" in {
```
Comment (dvoet) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/GoogleProjectRegistrationApiServiceSpec.scala:57]: add test that verifies a 404 response

File: core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/GoogleProjectRegistrationApiServiceSpec.scala [hunk 29]
```diff
@@ -0,0 +1,119 @@
...
+
+    Get(
+      "/googleProjects?billingProjectId=non-existent-billing-project&pageSize=10&offset=0"
+    ) ~> googleProjectRegistrationRoutes() ~> check {
+      status shouldEqual StatusCodes.OK
+      responseAs[Seq[GoogleProjectRegistration]] shouldEqual Seq.empty
+    }
+  }
+
+  it should "retrieve all Google projects if no billing project name is provided" in {
```
Comment (dvoet) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/GoogleProjectRegistrationApiServiceSpec.scala]: I am not convinced this adds value beyond the tests above. The only thing that it actually tests different is that the billing project name is optional.

**Review by dvoet — CHANGES_REQUESTED** [review]

File: core/src/main/resources/swagger/api-docs.yaml [hunk 30]
```diff
@@ -4658,6 +4658,47 @@ paths:
...
+      responses:
+        200:
+          description: Successful Request
+          content:
+            'application/json':
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/GoogleProjectResponse'
+        404:
```
Comment (dvoet) [inline core/src/main/resources/swagger/api-docs.yaml]: remove this, I don't think this api returns 404

**Review by dvoet — APPROVED** [review]
1 minor change then 👍

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/GoogleProjectRegistrationComponentSpec.scala [hunk 31]
```diff
@@ -53,4 +53,103 @@ class GoogleProjectRegistrationComponentSpec
...
+                                                                    GoogleProjectId("project2")
+                                                                ),
+                                                                None,
+                                                                10,
+                                                                0
+      )
+    )
+
+    result should have size 2
+    result.map(_.googleProjectId.value) should contain allOf ("project1", "project2")
```
Comment (davidangb) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/GoogleProjectRegistrationComponentSpec.scala]: use `only` instead of `allOf` for an accurate test:
```suggestion
    result.map(_.googleProjectId.value) should contain only ("project1", "project2")
```

`Seq("foo", "bar", "baz", "qux") should contain allOf ("foo", "bar")` passes, but `Seq("foo", "bar", "baz", "qux") should contain only ("foo", "bar")` fails.

(this applies to the rest of the test too)

**Review by davidangb — COMMENTED** [review]
I also have 1 minor change then 👍

**Review by davidangb — APPROVED** [review]
